### PR TITLE
warn user if they are upgrading with the default ingress gateway

### PR DESCRIPTION
**Pull request by ZhengzheYang**
install with `install_asm`
```
./install_asm --mode install -p zhengzhe-anthos-test -n test-cluster-1 -l us-central1-c
...
install_asm: *****************************
install_asm: Successfully installed ASM.

k get deployments -n istio-system                                                                
NAME                   READY   UP-TO-DATE   AVAILABLE   AGE
istio-ingressgateway   2/2     2            2           43s
istiod-asm-1104-6      2/2     2            2           6d19h
```

install with `asmcli` and choose to stop
```
2021-09-02T19:38:18.597013 asmcli: -------------
2021-09-02T19:38:19.083556 asmcli: [WARNING]: Defaults have changed between previous installation methods and this tool.
2021-09-02T19:38:19.149263 asmcli: [WARNING]: We detected an ASM ingress gateway currently running in the cluster that
2021-09-02T19:38:19.215977 asmcli: [WARNING]: would be disabled if continuing with default configuration.
2021-09-02T19:38:19.283483 asmcli: [WARNING]: If this is intended, please enter Y. If this is not intended, please enter
2021-09-02T19:38:19.348575 asmcli: [WARNING]: N and re-run the tool with the '--option legacy-default-ingressgateway'
Continue? [y/N] N
2021-09-02T19:38:37.844696 asmcli: [ERROR]: Stopping installation at user request.
```

install with `asmcli` and choose to continue
```
2021-09-02T19:39:46.788008 asmcli: -------------
2021-09-02T19:39:47.296882 asmcli: [WARNING]: Defaults have changed between previous installation methods and this tool.
2021-09-02T19:39:47.364806 asmcli: [WARNING]: We detected an ASM ingress gateway currently running in the cluster that
2021-09-02T19:39:47.434138 asmcli: [WARNING]: would be disabled if continuing with default configuration.
2021-09-02T19:39:47.499337 asmcli: [WARNING]: If this is intended, please enter Y. If this is not intended, please enter
2021-09-02T19:39:47.565495 asmcli: [WARNING]: N and re-run the tool with the '--option legacy-default-ingressgateway.'
Continue? [y/N] y
...
2021-09-02T19:41:08.631325 asmcli: -------------
✔ Istio core installed                                                                                                                                                                                                                                              
✔ Istiod installed                                                                                                                                                                                                                                                  
- Pruning removed resources                                                                                                                                                                                                                                           Removed HorizontalPodAutoscaler:istio-system:istio-ingressgateway.
  Removed PodDisruptionBudget:istio-system:istio-ingressgateway.
  Removed Deployment:istio-system:istio-ingressgateway.
  Removed Service:istio-system:istio-ingressgateway.
  Removed ServiceAccount:istio-system:istio-ingressgateway-service-account.
  Removed RoleBinding:istio-system:istio-ingressgateway-sds.
  Removed Role:istio-system:istio-ingressgateway-sds.
✔ Installation complete        
...
2021-09-02T19:41:26.573376 asmcli: *****************************
2021-09-02T19:41:26.639029 asmcli: Successfully installed ASM.
k get deployments -n istio-system           
NAME                READY   UP-TO-DATE   AVAILABLE   AGE
istiod-asm-1104-6   2/2     2            2           6d19h
```

install with `asmcli` and `--option legacy-default-ingressgateway` will not prompt the user

### CODE REVIEW

File: asmcli/lib/util.sh [hunk 1]
```diff
@@ -141,6 +141,16 @@ prompt_user_for_value() {
   fi
 }
 
+prompt() {
```
Comment (zerobfd) [inline asmcli/lib/util.sh:144]: What do you think about "prompt_default_no" just to make it a bit clearer?

File: asmcli/lib/validate.sh [hunk 2]
```diff
@@ -250,6 +251,26 @@ validate_ca_consistency() {
...
+  local INGRESS_GATEWAY
+  INGRESS_GATEWAY="$(kubectl get deployments -n istio-system -l operator.istio.io/component=IngressGateways)"
+  if [[ -z "${INGRESS_GATEWAY}" ]]; then
+    return
+  fi
+
+  local CUSTOM_OVERLAY; CUSTOM_OVERLAY="$(context_get-option "CUSTOM_OVERLAY")"
+  if [[ "${CUSTOM_OVERLAY}" =~ "legacy-default-ingressgateway" ]]; then
+    return
+  fi
```
Comment (zerobfd) [inline asmcli/lib/validate.sh:264]: This check is cheaper so it should probably go first.

File: asmcli/lib/validate.sh [hunk 3]
```diff
@@ -250,6 +251,26 @@ validate_ca_consistency() {
   fi
 }
 
+validate_no_ingress_gateway() {
+  local INGRESS_GATEWAY
+  INGRESS_GATEWAY="$(kubectl get deployments -n istio-system -l operator.istio.io/component=IngressGateways)"
+  if [[ -z "${INGRESS_GATEWAY}" ]]; then
+    return
+  fi
```
Comment (zerobfd) [inline asmcli/lib/validate.sh:259]: Here's what I would check:

1. there's a service called "istio-ingressgateway" in the "istio-system" namespace
2. there's a deployment of the same name in the same namespace

**Review by zerobfd — APPROVED** [review]

File: asmcli/lib/util.sh (hunk 1 above)
Comment (ZhengzheYang) [reply asmcli/lib/util.sh]: Done

File: asmcli/lib/validate.sh (hunk 2 above)
Comment (ZhengzheYang) [reply asmcli/lib/validate.sh]: Done

File: asmcli/lib/validate.sh (hunk 3 above)
Comment (ZhengzheYang) [reply asmcli/lib/validate.sh]: Done

**Pull request merged**